content/blog/r-mpi-part-two.md

```diff
@@ -0,0 +1,311 @@
+---
+author: "Wes Hinsley"
+date: 2023-07-12
+title: MPI from R - Part Two - Simple Comms
+best: false
+tags:
+ - R
+ - MPI
+ - HPC
+---
+
+Previously in [MPI From R](/blog/mpi-from-r-part-one-the-basics/) ... we 
+made an R package that could run code in multiple processes potentially
+spread across different computers (nodes), using MPI functions. Now 
+let's talk about sharing data between processes. We have a bit of structural
+work to do first.
+
+# Copying data between R and C++
+
+When R and C++ run togther, it's best to think of the two as separate
+worlds. We can't make a data structure that we can read from and write to 
+from both sides. Instead, we need a couple of helper functions below 
+(with thanks to Rich), to interface between the two. We'll define these in 
+a file `src/data.h` :-
+
+```
+#pragma once
+
+#include <cpp11.hpp>
+#include <vector>
+
+cpp11::sexp new_data(cpp11::doubles x);
+cpp11::sexp get_data(cpp11::external_pointer<std::vector<double>> ptr);
+```
+
+These two are called from R: the `new_data` function takes a
+R-flavoured vector of doubles, creates a copy of that vector in C++ world, and 
+gives back to R a _pointer_ to that new vector. R can do nothing with this 
+pointer except give it back to C++ later. The `get_data` function will copy
+the contents of the C++ vector back into an R-flavoured vector. Here's
+how we implement those two functions in `src/data.cpp` :-
+
+```
+#include "data.h"
+
+[[cpp11::register]]
+cpp11::sexp new_data(cpp11::doubles x) {
+  std::vector<double> *data = new std::vector<double>(x.size());
+  std::copy(x.begin(), x.end(), data->begin());
+  cpp11::external_pointer<std::vector<double>> ptr(data, true, false);
+  return cpp11::as_sexp(ptr);
+}
+
+[[cpp11::register]]
+cpp11::sexp get_data(cpp11::external_pointer<std::vector<double>> ptr) {
+  std::vector<double> *data = cpp11::as_cpp<cpp11::external_pointer<std::vector<double>>>(ptr).get();
+  cpp11::writable::doubles ret(data->size());
+  std::copy(data->begin(), data->end(), ret.begin());
+  return ret;
+}
+```
+
+# Adding a precision timer
+
+We're going to do some timing below, and the 
+[MPI_Wtime](https://www.mpich.org/static/docs/v3.3/www3/MPI_Wtime.html) 
+function gives us a double-precision[^1] number of seconds since some
+arbitrary origin.[^2] We'll wrap the MPI function as before, by inserting
+a line in the header `src/rmpi.h` :-
+
+```
+double get_mpi_time();
+```
+
+and the function itself in `src/rmpi.cpp` :-
+
+```
+[[cpp11::register]]
+double get_mpi_time() {
+  return MPI_Wtime();
+}
+  
+```
+
+# MPI Communication - Worst Case
+
+Now we're ready to write some very naive code, in which a number of
```

Comment:
I get lost through here working out what we're trying to achieve, I'm afraid.

Reply:
I think I've improved this a bit?